app/controllers/DataUploadController.scala

```diff
@@ -116,10 +128,60 @@ class DataUploadController @Inject()(sessionService: SessionService,
       )
   }
 
-  def process(res: List[UpscanCallback], empRef: String)
-             (hc: HeaderCarrier, schemeInfo: SchemeInfo, request: Request[_]): List[Future[(Int, Int)]] = {
-    for {
-      callbackData <- res
-    } yield fileProcessService.processCsvFile(callbackData, empRef)(hc, schemeInfo,request)
+  def processCsvFileDataFromFrontendV2(empRef:String): Action[JsValue] = authorisedActionWithBody(empRef) {
+    implicit request: Request[JsValue] =>
+      Logger.warn("receiving request on V2 processCsvFileDataFromFrontend")
+      val startTime =  System.currentTimeMillis()
+      request.body.validate[UpscanCsvFileData].fold(
+        valid = res => {
+          val schemeInfo: SchemeInfo = res.schemeInfo
+          Logger.debug("SCHEME TYPE: " + schemeInfo.schemeType)
+          deliverFileProcessingMetrics(startTime)
+
+          val processedFiles: List[Future[Either[Throwable, CsvFileSubmissions]]] = processCsvService.processFilesNew(res, streamFile)
+
+          val extractedSchemeData: Seq[Future[Either[Throwable, (Int, Int)]]] = processedFiles.map{ submission =>
+            submission.flatMap(processCsvService.extractSchemeDataNew(res.schemeInfo, empRef, _))}
+
+          Future.sequence(extractedSchemeData).flatMap{ oneFileResults => oneFileResults.find(_.isLeft) match {
+            case Some(Left(throwable: ERSFileProcessingException)) =>
+              Logger.error(s"[DataUploadController][processCsvFileDataFromFrontend] ERS file processing exception: ${throwable.message}")
+              deliverFileProcessingMetrics(startTime)
+              Future(Accepted(throwable.message))
+            case Some(Left(throwable)) =>
+              Logger.error(s"[DataUploadController][processCsvFileDataFromFrontend] Unknown exception: ${throwable.getMessage}, full exception: $throwable")
+              deliverFileProcessingMetrics(startTime)
+              Future(InternalServerError)
+            case None =>
+              val result: Seq[(Int, Int)] = oneFileResults.map(_.right.get)
+              val totalRowCount = result.foldLeft(0) ((accum,inputTuple) => accum + inputTuple._2)
+              sessionService.storeCallbackData(res.callbackData.head, totalRowCount).map {
+                case callback: Option[UpscanCallback] if callback.isDefined =>
+                  val numberOfSlices = result.map(_._1).sum
+                  Logger.error("Completed processing file")
+                  Ok(numberOfSlices.toString)
+                case _ =>
+                  Logger.error(
+                    s"[DataUploadController][processCsvFileDataFromFrontend] csv storeCallbackData failed" +
+                      s" while storing data, timestamp: ${java.time.LocalTime.now()}.")
+                  val exception = ERSFileProcessingException("csv callback data storage in sessioncache failed ", "Exception storing csv callback data")
```

Comment:
Is there much point in instantiating an ERSFileProcessingException here if we're not throwing it? Can we not just pass the String itself as a parameter for Accepted?

Reply:
good point - will change that